### PR TITLE
RFC

### PR DESCRIPTION
**Pull request by shibbirtanvin**

### CODE REVIEW

**Comment by shibbirtanvin** [comment]
I've already signed the Google Individual CLA according to the provided instructions. Thank you.

**Comment by fchollet** [comment]
Thanks for the PR. These are some interesting ideas, but ultimately this is way too prescriptive and high-level for Keras APIs. Having the framework restrict the user's decision for things like what activation and loss combinations to use, or how to normalize / standardize your data, is something that could work in a higher-level API closer to sklearn than Keras.

E.g. using crossentropy with relu is not a bug from Keras's perspective, it's a choice you're allowed to make. The space of possible actions is just way too large, covering way too many possible use cases, for Keras to be opinionated at that level of granularity.

**Comment by shibbirtanvin** [comment]
@fchollet : Thank you for your comments. We agree we should not restrict Keras users on how they use the library (choosing specific values of hyperparameter or normalized data values etc.). However, if we inform Keras users about their incorrect usage of certain combinations of parameter values, they would not be surprised by getting unexpected performance ( e.g., accuracy, loss related issues) from their models (some examples from Keras issues: [2975](https://github.com/keras-team/keras/issues/2975),  [15727](https://github.com/keras-team/keras/issues/15727),  [2711](https://github.com/keras-team/keras/issues/2711), [14729](https://github.com/keras-team/keras/issues/14729), [1917](https://github.com/keras-team/keras/issues/1917), [13006](https://github.com/keras-team/keras/issues/13006), [6259](https://github.com/keras-team/keras/issues/6259),  [5400](https://github.com/keras-team/keras/issues/5400), [6688](https://github.com/keras-team/keras/issues/6688)). Our technique would allow such contracts to be put in place to inform developers about possible incorrect usage and prescribe a possible solution to get expected performance from Keras APIs.

For example, when training a DNN with relu as the activation function in the hidden layers, the gradients of a large percentage of the neurons become zero and the training accuracy is low. We could prevent this prevalent problem by writing a contract on Fit API. To specify this contract, we use a variable 'zero_gradients_percentage' to capture such properties of the model during training. Thus, we can ensure that the training can happen without dying ReLu problem with restricted hidden layers' activation function obtaining expected accuracy from the model. This kind of training problem has appeared in the Keras issues ([3687](https://github.com/keras-team/keras/issues/3687), [7659](https://github.com/keras-team/keras/issues/7659)) as well. 
```python
1 @contract(context='hidden_layers', activation='str, !=relu', zero_gradients_percentage='float, <0.70')
2 def fit(self, x=None, y=None,...):
```
If annotated Fit API is called during training those models, the users will be notified of the following contract violation message to prevent this kind of mistake.
```
Contract violation for Model:fit(). zero_gradients_percentage of 12.07 caused dying relu problem, 
activation function should not be relu.
```
Similarly, we can specify contracts to prevent vanishing gradient, exploding gradient ([16057](https://github.com/keras-team/keras/issues/16057), [10847](https://github.com/keras-team/keras/issues/10847), [15912](https://github.com/keras-team/keras/issues/15912), [5564](https://github.com/keras-team/keras/issues/5564), [2711](https://github.com/keras-team/keras/issues/2711), [1244](https://github.com/keras-team/keras/issues/1244),  [2134](https://github.com/keras-team/keras/issues/2134)) training problems on Fit API.
 
Therefore, even though the solution is prescriptive and high-level in this PR, the design and implementation of our technique inform developers of a possible subset of mistakes of their model architecture, data properties, and training behavior along with the prescribed solution from DL experts. Eventually, this technique would facilitate and supplement Keras model debugging without the need for additional tools e.g., TensorBoard and Keras custom callbacks. 
 
Please let us know if you have more comments and questions.
